async_fuse/src/memfs/cache/mod.rs

```diff
@@ -0,0 +1,750 @@
+
+        let bucket_size = self.bucket_size_in_block.overflow_mul(self.block_size);
+        // The key in the hashmap, the end key is included
+        let start_key = offset.overflow_div(bucket_size);
+        let end_key = offset
+            .overflow_add(len)
+            .overflow_sub(1)
+            .overflow_div(bucket_size);
+        // The index in the bucket, the end index is included
+        let bucket_start_index = offset
+            .overflow_rem(bucket_size)
+            .overflow_div(self.block_size);
+        let bucket_end_index = offset
+            .overflow_add(len)
+            .overflow_sub(1)
+            .overflow_rem(bucket_size)
+            .overflow_div(self.block_size);
+        // The offset in the block, the end offset is included
+        let start_offset = offset.overflow_rem(self.block_size);
+        let end_offset = offset
+            .overflow_add(len)
+            .overflow_sub(1)
+            .overflow_rem(self.block_size);
+
+        debug!(
+            "start_key = {}, end_key = {}, bucket_start_index = {}, bucket_end_index = {}, start_offset = {}, end_offset = {}",
+            start_key, end_key, bucket_start_index, bucket_end_index, start_offset, end_offset,
+        );
+
+        for k in start_key..=end_key {
+            let start_index = if k == start_key {
+                bucket_start_index
+            } else {
+                0
+            };
+            let end_index = if k == end_key {
+                bucket_end_index.overflow_add(1)
+            } else {
+                self.bucket_size_in_block
+            };
+
+            match file_cache.get(&k, &guard) {
+                Some(bucket) => {
+                    self.queue
+                        .lock()
+                        .push(bucket.clone(), 0.overflow_sub(now_monotonic()));
+                    for (pos, block) in bucket
+                        .read()
+                        .get(start_index..end_index)
+                        .unwrap_or_else(|| {
+                            panic!(
+                                "error when getting range of {}..{} in the cache bucket",
+                                start_index, end_index
+                            )
+                        })
+                        .iter()
+                        .cloned()
+                        .enumerate()
+                    {
+                        let s = if k == start_key && pos == 0 {
+                            start_offset
+                        } else {
+                            0
+                        };
+
+                        let e = if k == end_key
+                            && pos == end_index.overflow_sub(start_index).overflow_sub(1)
+                        {
+                            end_offset.overflow_add(1)
+                        } else {
+                            self.block_size
+                        };
+
+                        result.push(IoMemBlock::new(block, s, e));
```

Comment:
Just want to check if my understanding is correct, this block can also be None, right? Since it may be deleted because of Cache capability.

Reply:
Yes, but we really check the data before the usage, it's guaranteed by the users, such as memfs.